pkg/proxy/proxy.go

```diff
@@ -62,19 +95,37 @@ func (pm *Manager) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// Determine which auth provider to use.
 	var provider string
+	if param := r.URL.Query().Get(ObotAuthProviderQueryParam); param != "" {
+		// If the provider is set in the query params, use that.
+		provider = param
+	} else if cookie, err := r.Cookie(ObotAccessTokenCookie); err == nil {
+		// Extract the provider from the cookie, if it's there.
+		cookieValue, err := url.QueryUnescape(cookie.Value)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to unescape cookie value: %v", err), http.StatusBadRequest)
+			return
+		}
 
-	if provider = r.URL.Query().Get(AuthProviderCookie); provider != "" {
-		// Set it as a cookie for the future.
-		http.SetCookie(w, &http.Cookie{
-			Name:  AuthProviderCookie,
-			Value: provider,
-			Path:  "/",
-		})
-	} else if c, err := r.Cookie(AuthProviderCookie); err == nil {
-		provider = c.Value
+		var contents CookieContents
+		if err = json.Unmarshal([]byte(cookieValue), &contents); err == nil {
+			provider = contents.AuthProvider
+
+			// Update the cookie to just be the token, which is what the auth provider expects.
+			cookie.Value = contents.Token
+			allCookies := r.Cookies()
+			r.Header.Del("Cookie")
+			for _, c := range allCookies {
+				if c.Name != ObotAccessTokenCookie {
+					r.AddCookie(c)
+				}
+			}
+			r.AddCookie(cookie)
+		}
```

Comment:
Looks like error is not handled when cookieValue is malformed.

Reply:
Yep, this also is intentional, so that we can have a smooth migration from the old way to the new way. We just leave the cookie alone if it is bad, and the user will have to reauthenticate anyway, which will replace it with a new, good cookie.